enhancements/dns/forward-dns-over-tls.md

```diff
@@ -0,0 +1,317 @@
+may be a need to forward requests to upstream DNS resolvers. Because DNS is
+unencrypted by default there are concerns in these environments around data
+privacy and integrity. Currently, there is no way to configure DNS-over-TLS for
+upstream resolvers so there is also no workaround to alleviate this security
+concern. The expansion of the `UpstreamResolver` and `ForwardPlugin` APIs  for 
+DNS request forwarding improves the current security posture for DNS on 
+OpenShift.
+
+### Goals
+
+1. A cluster administrator can specify a CA certificate or CA bundle, TLS configuration, and an upstream TLS Server Name for the default zone in the DNS configuration.
+2. A cluster administrator can specify a CA certificate or CA bundle, TLS configuration, and an upstream TLS Server Name for a non-default zone.
+3. A cluster administrator can opt-in to DNS-over-TLS (using the configured CA/bundle) for each listed upstream in the default zone.
+4. A cluster administrator can opt-in to DNS-over-TLS (using the configured CA/bundle) for a set of upstreams on a non-default zone.
+
+### Non-Goals
+
+This enhancement applies only to the CoreDNS Operator shipped as part of OpenShift. Anything outside that scope would be not applicable to this enhancement including the CoreDNS static pods managed by the `machine-config-operator`.
+
+## Proposal
+
+This proposal covers changes to two APIs. The first is a change to 
+`UpstreamResolvers`.  To enable a cluster administrator to configure 
+DNS-over-TLS, the `UpstreamResolvers` type will get two new optional members:
+
+`CABundle` references a ConfigMap that contains a single CA certificate 
+or a CA bundle.
+
+`ServerName` contains an FQDN to validate when establishing the connection. 
+This is relevant in particular with cases where a customer has multiple 
+upstreams behind one serving certificate (e.g. `1.1.1.1` and `2.2.2.2` 
+behind `dns.myorg.com`) as well as indicating what server we want to connect to.
+
+Since `UpstreamResolvers` doesn't allow for configuration of a custom zone, 
+it's assumed that the same TLS configuration will apply to _n_ `Upstream` 
+objects. 
+
+```go
+type UpstreamResolvers struct {
+  // ...<snip>
+
+  // caBundle references a ConfigMap that must contain either a single
+  // CA Certificate or a CA Bundle (in the case of multiple upstreams signed
+  // by different CAs). Populating a configmap name in this field
+  // effectively enables DNS-over-TLS. If this is empty or the configmap
+  // referenced does not exist, DNS-over-TLS will not be enabled.
+  //
+  // 1. The configmap must contain a `cabundle.crt` key
+  // 2. The value must be a PEM encoded certificate or certificate bundle.
+  // 3. The administrator must create this configmap in the openshift-config namespace.
+  // 4. The ServerName value must be supplied when CABundle is configured.
+  //
+  // +optional
+  // +kubebuilder:validation:Optional
+  CABundle v1.ConfigMapNameReference `json:"caBundle,omitempty"`
```

Comment:
this looks like the wrong level for this setting.  I choose servers inside the individual Upstreams right?

Reply:
`UpstreamResolvers` is only for the root zone so we're limited by what can be represented in the Corefile. The limitation in this case is that we can only have one server block for the root zone `.:5353`.